File: 2019/day15/day15.py
#!/usr/bin/env python3
# https://adventofcode.com/2019/day/15

import sys
from collections import deque
from pathlib import Path

sys.path.append(Path(__file__).parent.parent.as_posix())
from intcode.Intcode import Computer  # noqa

filename = sys.argv[1] if len(sys.argv) > 1 else "input.txt"
software = Path(filename).read_text()


droid = Computer()
droid.load(software)
droid.start(output_mode="yield")

NORTH = 1
SOUTH = 2
WEST = 3
EAST = 4

WALL = 0
EMPTY = 1
OXYGEN = 2

MOVES = [
    None,
    (0, 1),  # north
    (0, -1),  # south
    (-1, 0),  # west
    (1, 0),  # east
]


def bfs(droid, show):
    positions = {}
    x, y = 0, 0
    positions[(x, y)] = 4  # for S
    direction = EAST

    q = deque()
    q.append((0, 0, 0))
    seen = set()

    droids = {}
    droids[(0, 0)] = droid.clone()

    oxygen = None
    max_steps = 0

    # need to clone the droid state to avoid testing all directions and walk back to them
    # when there is a T or X junction
    while q:
        x, y, steps = q.popleft()

        max_steps = max(max_steps, steps)

        droid = droids[(x, y)]

        for direction in (NORTH, SOUTH, WEST, EAST):
            dx, dy = MOVES[direction]
            mx, my = x + dx, y + dy

            if (mx, my) in seen:
                continue
            seen.add((x, y))

            new_droid = droid.clone()
            new_droid.input.append(direction)
            state = new_droid.resume()
            assert state == "yield"
            status = new_droid.output.popleft()

            assert 0 <= status <= 2

            positions[(mx, my)] = status

            if status == OXYGEN:
                oxygen = (new_droid, steps + 1)
                q.clear()  # stop the bfs
                break

            if status != WALL:
                droids[(mx, my)] = new_droid
                q.append((mx, my, steps + 1))

    droids.clear()
    seen.clear()

    if show:
        inf = int(1e6)
        minx, maxx, miny, maxy = inf, -inf, inf, -inf
        for (x, y), _ in positions.items():
            minx = min(minx, x)
            maxx = max(maxx, x)
            miny = min(miny, y)
            maxy = max(maxy, y)

        lines = []
        for y in range(maxy, miny - 1, -1):
            row = ""
            for x in range(minx, maxx + 1):
                status = positions.get((x, y), 3)
                row += "# O?S"[status]
            lines.append(row)
        print("\n".join(lines))

    return oxygen, max_steps


(oxygen, steps), _ = bfs(droid, False)
print(steps)

_, distance = bfs(oxygen, False)
print(distance)

File: src/year2019/day15/day15.rs
//! [Day 15: Oxygen System](https://adventofcode.com/2019/day/15)

use rustc_hash::{FxHashMap, FxHashSet};
use std::collections::VecDeque;

use intcode::{Computer, State};

const NORTH: i64 = 1;
const SOUTH: i64 = 2;
const WEST: i64 = 3;
const EAST: i64 = 4;

const WALL: i64 = 0;
// const EMPTY: i64 = 1;
const OXYGEN: i64 = 2;

fn bfs(droid: &Computer) -> (Computer, i64, i64) {
    let mut q = VecDeque::new();
    let mut seen = FxHashSet::default();
    let mut droids = FxHashMap::default();

    q.push_back((0, 0, 0));
    droids.insert((0, 0), droid.clone());
    let mut max_steps = 0;

    while let Some((x, y, steps)) = q.pop_front() {
        max_steps = max_steps.max(steps);

        let droid = droids.get(&(x, y)).unwrap().clone();

        seen.insert((x, y));

        for direction in [NORTH, SOUTH, WEST, EAST] {
            let (dx, dy) = match direction {
                NORTH => (0, 1),
                SOUTH => (0, -1),
                WEST => (-1, 0),
                EAST => (1, 0),
                _ => unreachable!(),
            };

            let (mx, my) = (x + dx, y + dy);

            if seen.contains(&(mx, my)) {
                continue;
            }

            let mut new_droid = droid.clone();
            new_droid.push(direction);
            let State::Output(status) = new_droid.run() else {
                panic!("unattended state");
            };

            if status == OXYGEN {
                return (new_droid, steps + 1, 0);
            }

            if status != WALL {
                droids.insert((mx, my), new_droid);
                q.push_back((mx, my, steps + 1));
            }
        }
    }

    (Computer::new(), 0, max_steps)
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (i64, i64) {
    let droid = Computer::load(data);

    let (oxygen, steps, _) = bfs(&droid);
    let (_, _, distance) = bfs(&oxygen);

    (steps, distance)
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}
